servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -498,30 +526,95 @@ void markClosed() {
+                        this.closeAsync().subscribe();
+                    }
+                    break;
+                }
+            }
+        }
+
+        void markHealthy() {
+            // Marking healthy is generally called from a successful health check, after a connection was added.
+            // However, it is possible that in the meantime, the host entered an EXPIRED state, then ACTIVE, then failed
+            // to open connections and entered the UNHEALTHY state before the original thread continues execution here.
+            // In such case, the flipped state is not the same as the one that just succeeded to open a connection.
+            // In an unlikely scenario that the following connection attempts fail indefinitely, a health check task
+            // would leak and would not be cancelled. Therefore, we cancel it here and allow failures to trigger a new
+            // health check.
+            Object oldState = connStateUpdater.getAndUpdate(this, previous -> {
+                if (previous.state.getClass().equals(HealthCheck.class)) {
+                    return new ConnState(previous.connections, STATE_ACTIVE_NO_FAILURES);
+                }
+                return previous;
+            }).state;
+            cancelIfHealthCheck(oldState);
+        }
+
+        void markUnhealthy(ConnectionFactory<Addr, ? extends C> connectionFactory) {
+            assert healthCheckConfig != null;
+            for (;;) {
+                ConnState previous = connStateUpdater.get(this);
+
+                if (!ActiveState.class.equals(previous.state.getClass()) || previous.connections.length > 0) {
+                    break;
+                }
+
+                ActiveState previousState = (ActiveState) previous.state;
+                if (previousState.failedConnections + 1 < this.healthCheckConfig.failedThreshold) {
+                    final ActiveState nextState = previousState.forNextFailedConnection();
+                    if (connStateUpdater.compareAndSet(this, previous,
+                            new ConnState(previous.connections, nextState))) {
+                        LOGGER.debug("Active host for address {} failed to open {} connections" +
+                                        " ({} consecutive failures trigger health check).",
+                                address, nextState.failedConnections, healthCheckConfig.failedThreshold);
+                        break;
+                    }
+                    // another thread won the race, try again
+                    continue;
+                }
+
+                final HealthCheck<Addr, C> healthCheck = new HealthCheck<>(connectionFactory, this);
+                final ConnState nextState = new ConnState(previous.connections, healthCheck);
+                if (connStateUpdater.compareAndSet(this, previous, nextState)) {
+                    LOGGER.debug("Triggering health check for address {}", address);
```

Comment:
Consider: Triggering health check for address {} after {} failed attempts to open a new connection